### ISSUE TITLE
PathErrorKraken when trying to make an absolute path absolute

### ISSUE THREAD

**Issue by raptor-xxl** [KSP-CKAN/CKAN#1940]
Thanks for wanting to bring an issue to our attention!

Before you open a ticket, please do a quick search of the existing issues to see if it's already been brought up.

**CKAN Version:** v1.20.1

**Operating System:** Win 8.1 x64

**The issue you are experiencing:** Unhandled exception

**How to recreate this issue:** Try to upgrade Dang it! Continued mod

**CKAN error codes (if applicable):** 

````
CKAN.PathErrorKraken: c:/program files (x86)/Steam/SteamApps/common/Kerbal Space Program/GameData/DangIt/ModuleManager is already absolute
   at CKAN.KSPPathUtils.ToAbsolute(String path, String root)
   at CKAN.ModuleInstaller.<AddParentDirectories>c__AnonStorey0.<>m__0(String dir)
   at System.Linq.Enumerable.<SelectManyIterator>d__1`2.MoveNext()
   at System.Linq.Enumerable.WhereEnumerableIterator`1.MoveNext()
   at System.Collections.Generic.HashSet`1.UnionWith(IEnumerable`1 other)
   at System.Collections.Generic.HashSet`1..ctor(IEnumerable`1 collection, IEqualityComparer`1 comparer)
   at CKAN.ModuleInstaller.AddParentDirectories(HashSet`1 directories)
   at CKAN.ModuleInstaller.Uninstall(String modName)
   at CKAN.ModuleInstaller.AddRemove(IEnumerable`1 add, IEnumerable`1 remove, Boolean enforceConsistency)
   at CKAN.Main.WasSuccessful(Action action)
   at CKAN.Main.InstallMods(Object sender, DoWorkEventArgs e)
   at System.ComponentModel.BackgroundWorker.OnDoWork(DoWorkEventArgs e)
   at System.ComponentModel.BackgroundWorker.WorkerThreadStart(Object argument)
````

**Comment by politas** [comment]
We've been getting a few reports of this issue. I think we need to go over the logic for the parent directories deletion code.

**Comment by stmobo** [comment]
I've been having the same issue as the OP (trying to update DangItContinued), and I've been looking at the code to maybe try and help debug. I think I've figured out what the problem is.

In short: the code is stumbling on conflicting capitalization of paths. Even though paths are supposed to be case-insensitive (in NTFS, anyways) the code is using normal, case-sensitive, string comparisons.

The problem lies in [line 929 of ModuleInstaller.cs](../blob/master/Core/ModuleInstaller.cs#L929).
The given path to AddParentDirectories begins with "c:/program files (x86)/_Steam_".
![Looking at the top window bar,](https://cloud.githubusercontent.com/assets/1782189/20548136/7a054b60-b0e6-11e6-8dca-545f4a17fc71.PNG), the given gameDir seems to be "c:/program files (x86)/_steam_". Note the differing capitalizations.

This is probably causing the comparison to fail here, passing the path on to ToAbsolute-- and triggering the crash.

You _could_ fix this particular instance by lower-casing the paths in this comparison, but that seems really hacky-- there might be other issues with case like this elsewhere.

**Comment by politas** [comment]
Paths are not case-insensitive on Linux or MacOS.

**Comment by stmobo** [comment]
Yeah, I was afraid of that. It'll have to be dealt with in Windows, though.

**Comment by politas** [comment]
So, it sounds like we're getting different case results interrogating the filesystem at different times. Unless the KSP instances have been added from the command line and the incorrect capitalisation was typed in?

**Comment by stmobo** [comment]
I dunno about the OP, but I haven't done anything like that. I think I did manually select a KSP install, but that was through the GUI, not the command line.

**Comment by politas** [comment]
Can you have a look in your registry.json at the "installed_files" for the mod that's getting the error? The file paths should start with "GameData/[...]"

**Comment by stmobo** [comment]
They do indeed begin with "GameData/DangIt".

**Comment by politas** [comment]
If you could also look in the registry "CurrentUser/software/ckan" should have the list of KSP installs - I'm assuming you have the capitalised "Steam" there?

**Comment by stmobo** [comment]
Nope, "KSPInstancePath_0" has uncapitalized "steam" in the path.

**Comment by politas** [comment]
Ok. And if you open up a command prompt and use tab-autocomplete to navigate to that directory, what is the case?

**Comment by stmobo** [comment]
Beginning from C:\, tab-autocomplete takes me to "'Program Files (x86)'", "Steam", "steamapps", "common", and "Kerbal Space Program". 
![cmd](https://cloud.githubusercontent.com/assets/1782189/20549082/c770c11a-b0ee-11e6-8f78-347ecc88cda5.PNG)

**Comment by politas** [comment]
Ok. Can you edit the path in the registry to match the actual capitalisation, and then test?

**Comment by politas** [comment]
Windows is weird when it comes to filesystem case. Files and directories all actually have specific capitalisation, but for historic reasons, Windows file operations are often, but not always, handled in case-insensitive ways. It means that you can have directories you can't effectively access, if you manage to get two directories called (for instance)

    C:\test\
    C:\Test\

Some apps will not even see one of those directories.

**Comment by stmobo** [comment]
Just did that (I copied the path from an Explorer window open to the game directory into regedit) and it seems to have worked without a hitch.

**Comment by politas** [comment]
Excellent! Looks like we maybe need to intermittently make sure our paths in the registry match the correct capitalisation from the filesystem. Don't know why it only crapped out on the one mod, though.

**Comment by stmobo** [comment]
About that: I remember also having an issue a couple days (~1-2 weeks? Same KSP version at least) back while trying to update "EVA Parachutes and Ejection Seats"-- I didn't think to investigate the issue. From what I remember I just deleted it from GameData and had CKAN install the new version.

**Comment by politas** [comment]
Most issues with CKAN can be fixed by uninstalling and reinstalling mods.

**Comment by macdjord** [comment]
I'm getting this same bug while updating 'EVA Parachutes & Ejection Seats' from 0.1.11 to 0.1.12.

I have to ask - who in the world writes a convert-to-absolute-path function that _raises a fatal error_ if the path is already absolute instead of just returning the path unchanged?

CKAN Version: v1.22.1-0-g17427a (beta)
System: Windows 10 (64 bit)

Error message:
```
See the end of this message for details on invoking 
just-in-time (JIT) debugging instead of this dialog box.

************** Exception Text **************
CKAN.PathErrorKraken: c:/program files (x86)/Steam/SteamApps/common/Kerbal Space Program/GameData/VanguardTechnologies/Parts/EjModule2 is already absolute
   at CKAN.KSPPathUtils.ToAbsolute(String path, String root)
   at CKAN.ModuleInstaller.<AddParentDirectories>c__AnonStorey0.<>m__0(String dir)
   at System.Linq.Enumerable.<SelectManyIterator>d__16`2.MoveNext()
   at System.Linq.Enumerable.WhereEnumerableIterator`1.MoveNext()
   at System.Collections.Generic.HashSet`1.UnionWith(IEnumerable`1 other)
   at System.Collections.Generic.HashSet`1..ctor(IEnumerable`1 collection, IEqualityComparer`1 comparer)
   at CKAN.ModuleInstaller.AddParentDirectories(HashSet`1 directories)
   at CKAN.ModuleInstaller.Uninstall(String modName)
   at CKAN.ModuleInstaller.AddRemove(IEnumerable`1 add, IEnumerable`1 remove, Boolean enforceConsistency)
   at CKAN.ModuleInstaller.Upgrade(IEnumerable`1 modules, NetAsyncModulesDownloader netAsyncDownloader, Boolean enforceConsistency)
   at CKAN.Main.WasSuccessful(Action action)
   at CKAN.Main.InstallMods(Object sender, DoWorkEventArgs e)
   at System.ComponentModel.BackgroundWorker.OnDoWork(DoWorkEventArgs e)
   at System.ComponentModel.BackgroundWorker.WorkerThreadStart(Object argument)


************** Loaded Assemblies **************
mscorlib
    Assembly Version: 4.0.0.0
    Win32 Version: 4.6.1586.0 built by: NETFXREL2
    CodeBase: file:///C:/Windows/Microsoft.NET/Framework64/v4.0.30319/mscorlib.dll
----------------------------------------
ckan
    Assembly Version: 0.0.0.0
    Win32 Version: 0.0.0.0
    CodeBase: file:///C:/Users/macdjord/Desktop/ckan.exe
----------------------------------------
System.Configuration
    Assembly Version: 4.0.0.0
    Win32 Version: 4.6.1586.0 built by: NETFXREL2
    CodeBase: file:///C:/WINDOWS/Microsoft.Net/assembly/GAC_MSIL/System.Configuration/v4.0_4.0.0.0__b03f5f7f11d50a3a/System.Configuration.dll
----------------------------------------
System
    Assembly Version: 4.0.0.0
    Win32 Version: 4.6.1586.0 built by: NETFXREL2
    CodeBase: file:///C:/WINDOWS/Microsoft.Net/assembly/GAC_MSIL/System/v4.0_4.0.0.0__b77a5c561934e089/System.dll
----------------------------------------
System.Core
    Assembly Version: 4.0.0.0
    Win32 Version: 4.6.1586.0 built by: NETFXREL2
    CodeBase: file:///C:/WINDOWS/Microsoft.Net/assembly/GAC_MSIL/System.Core/v4.0_4.0.0.0__b77a5c561934e089/System.Core.dll
----------------------------------------
System.Xml
    Assembly Version: 4.0.0.0
    Win32 Version: 4.6.1586.0 built by: NETFXREL2
    CodeBase: file:///C:/WINDOWS/Microsoft.Net/assembly/GAC_MSIL/System.Xml/v4.0_4.0.0.0__b77a5c561934e089/System.Xml.dll
----------------------------------------
System.Windows.Forms
    Assembly Version: 4.0.0.0
    Win32 Version: 4.6.1586.0 built by: NETFXREL2
    CodeBase: file:///C:/WINDOWS/Microsoft.Net/assembly/GAC_MSIL/System.Windows.Forms/v4.0_4.0.0.0__b77a5c561934e089/System.Windows.Forms.dll
----------------------------------------
System.Drawing
    Assembly Version: 4.0.0.0
    Win32 Version: 4.6.1586.0 built by: NETFXREL2
    CodeBase: file:///C:/WINDOWS/Microsoft.Net/assembly/GAC_MSIL/System.Drawing/v4.0_4.0.0.0__b03f5f7f11d50a3a/System.Drawing.dll
----------------------------------------
System.Transactions
    Assembly Version: 4.0.0.0
    Win32 Version: 4.6.1586.0 built by: NETFXREL2
    CodeBase: file:///C:/WINDOWS/Microsoft.Net/assembly/GAC_64/System.Transactions/v4.0_4.0.0.0__b77a5c561934e089/System.Transactions.dll
----------------------------------------
Microsoft.GeneratedCode
    Assembly Version: 1.0.0.0
    Win32 Version: 4.6.1586.0 built by: NETFXREL2
    CodeBase: file:///C:/WINDOWS/Microsoft.Net/assembly/GAC_MSIL/System.Xml/v4.0_4.0.0.0__b77a5c561934e089/System.Xml.dll
----------------------------------------
System.Numerics
    Assembly Version: 4.0.0.0
    Win32 Version: 4.6.1586.0 built by: NETFXREL2
    CodeBase: file:///C:/WINDOWS/Microsoft.Net/assembly/GAC_MSIL/System.Numerics/v4.0_4.0.0.0__b77a5c561934e089/System.Numerics.dll
----------------------------------------
System.Runtime.Serialization
    Assembly Version: 4.0.0.0
    Win32 Version: 4.6.1586.0 built by: NETFXREL2
    CodeBase: file:///C:/WINDOWS/Microsoft.Net/assembly/GAC_MSIL/System.Runtime.Serialization/v4.0_4.0.0.0__b77a5c561934e089/System.Runtime.Serialization.dll
----------------------------------------
System.Xml.Linq
    Assembly Version: 4.0.0.0
    Win32 Version: 4.6.1586.0 built by: NETFXREL2
    CodeBase: file:///C:/WINDOWS/Microsoft.Net/assembly/GAC_MSIL/System.Xml.Linq/v4.0_4.0.0.0__b77a5c561934e089/System.Xml.Linq.dll
----------------------------------------
System.Data
    Assembly Version: 4.0.0.0
    Win32 Version: 4.6.1636.0 built by: NETFXREL3STAGE
    CodeBase: file:///C:/WINDOWS/Microsoft.Net/assembly/GAC_64/System.Data/v4.0_4.0.0.0__b77a5c561934e089/System.Data.dll
----------------------------------------
Anonymously Hosted DynamicMethods Assembly
    Assembly Version: 0.0.0.0
    Win32 Version: 4.6.1586.0 built by: NETFXREL2
    CodeBase: file:///C:/WINDOWS/Microsoft.Net/assembly/GAC_64/mscorlib/v4.0_4.0.0.0__b77a5c561934e089/mscorlib.dll
----------------------------------------
Microsoft.CSharp
    Assembly Version: 4.0.0.0
    Win32 Version: 4.6.1586.0
    CodeBase: file:///C:/WINDOWS/Microsoft.Net/assembly/GAC_MSIL/Microsoft.CSharp/v4.0_4.0.0.0__b03f5f7f11d50a3a/Microsoft.CSharp.dll
----------------------------------------
System.Dynamic
    Assembly Version: 4.0.0.0
    Win32 Version: 4.6.1586.0
    CodeBase: file:///C:/WINDOWS/Microsoft.Net/assembly/GAC_MSIL/System.Dynamic/v4.0_4.0.0.0__b03f5f7f11d50a3a/System.Dynamic.dll
----------------------------------------
System.EnterpriseServices
    Assembly Version: 4.0.0.0
    Win32 Version: 4.6.1586.0 built by: NETFXREL2
    CodeBase: file:///C:/WINDOWS/Microsoft.Net/assembly/GAC_64/System.EnterpriseServices/v4.0_4.0.0.0__b03f5f7f11d50a3a/System.EnterpriseServices.dll
----------------------------------------
Accessibility
    Assembly Version: 4.0.0.0
    Win32 Version: 4.6.1586.0 built by: NETFXREL2
    CodeBase: file:///C:/WINDOWS/Microsoft.Net/assembly/GAC_MSIL/Accessibility/v4.0_4.0.0.0__b03f5f7f11d50a3a/Accessibility.dll
----------------------------------------

************** JIT Debugging **************
To enable just-in-time (JIT) debugging, the .config file for this
application or computer (machine.config) must have the
jitDebugging value set in the system.windows.forms section.
The application must also be compiled with debugging
enabled.

For example:

<configuration>
    <system.windows.forms jitDebugging="true" />
</configuration>

When JIT debugging is enabled, any unhandled exception
will be sent to the JIT debugger registered on the computer
rather than be handled by this dialog box.
```

**Comment by macdjord** [comment]
Same thing happens trying to uninstall it. Deleting the folder in GameData manually, then uninstalling & reinstalling it with CKAN, worked.

**Comment by tjanke42** [comment]
Same issue when updating Smart Parts.  I'll try the workaround **macdjord** suggested.

**Comment by occar421** [comment]
I just got this error.
It was solved by add same config on 'Select KSP instal' & remove previous (auto generated) config.

**Comment by ashein** [comment]
Just got it yesterday when updating `EVA Enhancements Continued`, a day before auto-updating existing mods worked fine, nothing changed as far as I remember. Tried the usual "remove the mod + reinstall the mod", but apparently the "path is already absolute" triggered on mod removal before update, not on the update itself. That was on Win10. Manually correcting `KSPInstancePath_0` in the registry helped me.

**Comment by HebaruSan** [comment]
> I have to ask - who in the world writes a convert-to-absolute-path function that raises a fatal error if the path is already absolute instead of just returning the path unchanged?

That would indeed be pretty silly, but the function is actually misleadingly named. It's not trying to convert a relative path to an absolute path; it's trying to _combine_ a relative path and an absolute path into one longer absolute path. The fatal error makes sense because you don't want to end up with something like:

- C:\Program Files (x86)\Steam\C:\Kerbal Space Program\GameData

**Comment by HebaruSan** [comment]
Fixed in #2122; closing.
